### PR TITLE
[DT][GPU] Implement GPUEncodingLayoutAttr.

### PR DESCRIPTION
**Pull request by hanhanW**
The revision introduces GPUEncodingLayoutAttr and moves the encoding materialization logic to the attribute implementation using external model mechanism. I.e., the implementation is in Codegen/ExternalInterfaces/GPUEncodingExternalModels.cpp. They are implemented through external model mechanism because we do not want to put burdens to the dialect, and we can have better code structure. They are domain-specific, but not IREEGPU dialect IRs.

There are no functional changes on the moved codes. The additional changes are:

- Remove IREE::GPU namespaces because they are under the same
namespace now.
- Create lowerContractionOpToMultiMmaOp method, which lowers linalg
contraction ops to multi_mma ops. The code is moved from
- GPUMaterializationEncoding.cpp. It returns nullptr instead if there is
a failure.

No new tests because it does not change the functionality; it is covered by gpu_materialize_encoding.mlir.

### CODE REVIEW

**Comment by hanhanW** [comment]
This is simliar to https://github.com/iree-org/iree/pull/19358 but put the domain-specific logic to Codegen/ExternalInterfaces.

@bjacob what do you think about this? Does it look good?

File: compiler/src/iree/compiler/Codegen/ExternalInterfaces/GPUEncodingExternalModels.cpp [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+
+static Operation *lowerContractionOpToMultiMmaOp(OpBuilder &builder,
+                                                 linalg::LinalgOp linalgOp,
+                                                 ValueRange operands,
+                                                 TargetAttr targetAttr) {
+  if (!linalgOp.hasPureTensorSemantics()) {
+    return nullptr;
+  }
+  FailureOr<linalg::ContractionDimensions> contractionDims =
+      linalg::inferContractionDims(linalgOp);
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/ExternalInterfaces/GPUEncodingExternalModels.cpp:209]: Only inferring constractionDims is not always enough to confirm that an op is a contraction. It is best to check `linalg::isaContractionOpInterface(linalgOp)` before this, which checks the body of the operation as well.

I don't think we will get any non-contraction ops with encodings anyway, so it is sort of just a nit, but I think it is best to be safe here.

**Review by Max191 — APPROVED** [review]
Cool! It's nice having all the layout logic in one place with the external interfaces.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td [hunk 2]
```diff
@@ -314,6 +314,29 @@ def IREEGPU_MmaScheduleAttr : AttrDef<IREEGPU_Dialect, "MMASchedule"> {
   let assemblyFormat = "`<` struct(params) `>`";
 }
 
+//===----------------------------------------------------------------------===//
+// iree_gpu.gpu_encoding_layout
+//===----------------------------------------------------------------------===//
+
+// This attribute needs to implement the Codegen::LayoutAttrInterface and
+// Encoding::EncodingLayoutAttrInterface. They are implemented through external
+// model mechanism because we do not want to put burdens to the dialect, and we
```
Comment (bjacob) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td]: We do not want to relocate domain-specific logic to the dialect implementation

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td [hunk 3]
```diff
@@ -314,6 +314,29 @@ def IREEGPU_MmaScheduleAttr : AttrDef<IREEGPU_Dialect, "MMASchedule"> {
...
+// can have better code structure. See the implementations in
+// compiler/Codegen/ExternalInterfaces.
+def IREEGPU_GPUEncodingLayoutAttr :
+    AttrDef<IREEGPU_Dialect, "GPUEncodingLayout", [
+]> {
+  let mnemonic = "gpu_encoding_layout";
+  let summary = "The encoding layout attribute for VMVX backend";
```
Comment (bjacob) [inline compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td]: Copy and paste? This isn't VMVX specific.

**Review by bjacob — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/ExternalInterfaces/GPUEncodingExternalModels.cpp (hunk 1 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/ExternalInterfaces/GPUEncodingExternalModels.cpp:209]: Wow, that's surprising. I thought that it has an assertion. Let me add a check.

File: compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td (hunk 3 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td]: Nice catch! I checked it few times and I did not catch it.

**Comment by hanhanW** [comment]
I fixed few more copy-paste errors and move the comment of the attribute to its description section. It is ready to go, I'll land it once the CI is green.

**Pull request merged**